mrc-2446 Update Phase Rt values from Phase Editor

EmmaLRussell opened the pull request:
Allows user to update Rt values of Phases. Rt values are constrained by min 0 and max 4, and are only saved with 2 decimal places. 

Uses [this technique](https://michaelnthiessen.com/force-re-render/) (update element key) to ensure the input is correctly re-rendered if the user enters invalid value. This silently updates the value to a valid one - even though there is a note in the dialog header, perhaps there should be more feedback to user that their value was not valid?

EmmaLRussell committed:
Allow updates of Rt values in phase editor

EmmaLRussell committed:
Merge branch 'main' into mrc-2446

EmmaLRussell committed:
Added tests

EmmaLRussell committed:
prevent unwanted text select on drag phas

EmmaLRussell commented on src/app/static/src/components/parameters/EditPhases.vue:
```diff
@@ -2,27 +2,32 @@
+                 @mouseup="mouseUp">
+              <div class="slider-spike" :class="phaseClassFromIndex(index+1)"></div>
+              <div class="slider-text">
+                <div @mousedown.prevent="">
```
This prevents unintended and distracting selection of text when dragging phases...

EmmaLRussell commented on src/app/static/src/components/parameters/EditPhases.vue:
```diff
@@ -32,12 +37,23 @@
+                    :value="sliderValues[index].value.rt"
+                    step="0.01"
+                    @change="updateRt(index, $event)"
+                    @mousedown.stop="">
```
...and this prevents unintended dragging of phases when user selects text with drag in Rt input

EmmaLRussell committed:
lint

mwoodbri reviewed (approved):
Wow... those are some tricky edge cases to deal with but it works really well.

I only had a couple of observations - maybe the curse of Firefox and certainly shouldn't block progress. Demonstrated by this screenshot:

![rt](https://user-images.githubusercontent.com/1724545/122246964-e5d54700-cebe-11eb-8ce4-3ad5bf1d7949.png)

So obviously the Rt input is wider than the shaded box.. but its siblings are too. It just seems to be the parent (i.e. `div.slider`) that is narrower. So if you wanted to fix that then I guess you could shrink the input or widen the box.

And a really minor thing - if they're overlapping then you can't bring the lower one forward by clicking in the input, you have to click in the shaded box instead.

EmmaLRussell commented:
>So obviously the Rt input is wider than the shaded box.. but its siblings are too. It just seems to be the parent (i.e. `div.slider`) that is narrower. So if you wanted to fix that then I guess you could shrink the input or widen the box.
> 

Ha how weird, the width (again) works-on-my-machine (famous last words), though I possibly just tested on Chrome this time. I'll try pinning down the width of the text divs and input and see if that helps, but it is strange that you're seeing such different behaviour. 

> And a really minor thing - if they're overlapping then you can't bring the lower one forward by clicking in the input, you have to click in the shaded box instead.

Good point!

mwoodbri commented:
> Ha how weird, the width (again) works-on-my-machine (famous last words), though I possibly just tested on Chrome this time. I'll try pinning down the width of the text divs and input and see if that helps, but it is strange that you're seeing such different behaviour.

A fix/hack: `<input type="number" ... style="width: 100%">`

Nicolas-Dolan commented on src/app/static/src/components/parameters/EditPhases.vue:
```diff
@@ -2,27 +2,32 @@
-                <div class="slider-text">
+      <p>
+        Click on a Phase to drag it to a new start date.
+        Rt values must be between {{rtMin}} and {{rtMax}}.
```
Making this dynamic was a nice touch!

Nicolas-Dolan reviewed (approved):
Works well. You could have a small message pop up when a user enters an invalid value but would be a little redundant in my opinion as the required range is stated just above. 

If you wanted to get fancy you could add some visual cues to emphasise when the value is corrected like a flash on the input or a spinner that appears for a second next to the input (as if it is working out an appropriate value) or have the range message above flash red momentarily.

EmmaLRussell committed:
Fix input widths, bring slider to front on click input, show animation when correct out of range rt

EmmaLRussell committed:
Fix bug where values lost when click away from slider, add tests

EmmaLRussell commented:
I've implemented both your suggestions:
- I've fixed the width of the inputs, hopefully that should fix the overflow Mark was seeing. (Still confused that I'm not seeing the same styling that you are!)
- Clicking on input will bring that slider to the front - NB 'click' event rather than 'mousedown'. User cannot drag slider by mousedowning on the input, they need to complete the click to interact. 
- I've gone with Nick's suggestion of putting a little bit of animation in to highlight the automatic resetting of values which are out of range. I've used [animate.css](https://animate.style/) to put a little shake on both the input in question and the text describing the allowed range. See what you think.

mwoodbri commented:
> I've implemented both your suggestions:
> 
>     * I've fixed the width of the inputs, hopefully that should fix the overflow Mark was seeing. (Still confused that I'm not seeing the same styling that you are!)

It does indeed avoid the overflow, thanks

>     * Clicking on input will bring that slider to the front - NB 'click' event rather than 'mousedown'. User cannot drag slider by mousedowning on the input, they need to complete the click to interact.

Works for me

>     * I've gone with Nick's suggestion of putting a little bit of animation in to highlight the automatic resetting of values which are out of range. I've used [animate.css](https://animate.style/) to put a little shake on both the input in question and the text describing the allowed range. See what you think.

Enjoying the wobble
